### ISSUE TITLE
Failed to fuse (linalg.generic -> linalg.tensor_reshape -> linalg.generic)

### ISSUE THREAD

**Issue by asaadaldien** [google/iree#3302]
FusionOfTensorOpsPass failing to fuse the following ops into a single `linalg.genric`
```
module {
  func @serving_default_ex_dispatch_14() {
    %c0 = constant 0 : index
    %0 = hal.interface.load.tensor @legacy_io::@arg1, offset = %c0 : tensor<128xf32>
    %1 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d1)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = ["parallel", "parallel"]} ins(%0 : tensor<128xf32>) {
    ^bb0(%arg0: f32):  // no predecessors
      linalg.yield %arg0 : f32
    } -> tensor<384x128xf32>
    %2 = hal.interface.load.tensor @legacy_io::@arg0, offset = %c0 : tensor<384x4x32xf32>
    %3 = linalg.tensor_reshape %1 [affine_map<(d0, d1, d2) -> (d0)>, affine_map<(d0, d1, d2) -> (d1, d2)>] : tensor<384x128xf32> into tensor<384x4x32xf32>
    %4 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d0, d2)>, affine_map<(d0, d1, d2) -> (d1, d0, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%2, %3 : tensor<384x4x32xf32>, tensor<384x4x32xf32>) {
    ^bb0(%arg0: f32, %arg1: f32):  // no predecessors
      %5 = addf %arg0, %arg1 : f32
      linalg.yield %5 : f32
    } -> tensor<4x384x32xf32>
    hal.interface.store.tensor %4, @legacy_io::@ret0, offset = %c0 : tensor<4x384x32xf32>
    return
  }
  hal.interface @legacy_io attributes {sym_visibility = "private"} {
    hal.interface.binding @arg0, set=0, binding=0, type="StorageBuffer", access="Read"
    hal.interface.binding @arg1, set=0, binding=1, type="StorageBuffer", access="Read"
    hal.interface.binding @ret0, set=0, binding=2, type="StorageBuffer", access="Write|Discard"
  }
}
```

**Comment by asaadaldien** [comment]
Before the pass. The problem is fusing `linalg.tenso_reshape` with `linalg.generic` that represents a transpose 
```
module {
  func @serving_default_ex_dispatch_14() {
    %c0 = constant 0 : index
    %0 = hal.interface.load.tensor @legacy_io::@arg0, offset = %c0 : tensor<384x128xf32>
    %1 = hal.interface.load.tensor @legacy_io::@arg1, offset = %c0 : tensor<128xf32>
    %2 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d1)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = ["parallel", "parallel"]} ins(%1 : tensor<128xf32>) {
    ^bb0(%arg0: f32):  // no predecessors
      linalg.yield %arg0 : f32
    } -> tensor<384x128xf32>
    %3 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = ["parallel", "parallel"]} ins(%0, %2 : tensor<384x128xf32>, tensor<384x128xf32>) {
    ^bb0(%arg0: f32, %arg1: f32):  // no predecessors
      %6 = addf %arg0, %arg1 : f32
      linalg.yield %6 : f32
    } -> tensor<384x128xf32>
    %4 = linalg.tensor_reshape %3 [affine_map<(d0, d1, d2) -> (d0)>, affine_map<(d0, d1, d2) -> (d1, d2)>] : tensor<384x128xf32> into tensor<384x4x32xf32>
    %5 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d0, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%4 : tensor<384x4x32xf32>) {
    ^bb0(%arg0: f32):  // no predecessors
      linalg.yield %arg0 : f32
    } -> tensor<4x384x32xf32>
    hal.interface.store.tensor %5, @legacy_io::@ret0, offset = %c0 : tensor<4x384x32xf32>
    return
  }
  hal.interface @legacy_io attributes {sym_visibility = "private"} {
    hal.interface.binding @arg0, set=0, binding=0, type="StorageBuffer", access="Read"
    hal.interface.binding @arg1, set=0, binding=1, type="StorageBuffer", access="Read"
    hal.interface.binding @ret0, set=0, binding=2, type="StorageBuffer", access="Write|Discard"
  }
}
```

**Comment by asaadaldien** [comment]
The following two ops could be canonicalized: 
```
    %4 = linalg.tensor_reshape %3 [affine_map<(d0, d1, d2) -> (d0)>, affine_map<(d0, d1, d2) -> (d1, d2)>] : tensor<384x128xf32> into tensor<384x4x32xf32>
    %5 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d0, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%4 : tensor<384x4x32xf32>) {
    ^bb0(%arg0: f32):  // no predecessors
      linalg.yield %arg0 : f32
    } -> tensor<4x384x32xf32>
```

```
    %5 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d0 * 32 + d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%4 : tensor<384x128xf32>) {
    ^bb0(%arg0: f32):  // no predecessors
      linalg.yield %arg0 : f32
    } -> tensor<4x384x32xf32>
```

Which is simply by applying the permutation map on the reassociations resulting   
```
[affine_map<(d0, d1, d2) -> (d1)>, affine_map<(d0, d1, d2) -> (d0, d2)>]
```
then linearize it:
```
affine_map<(d0, d1, d2) -> (d1, d0 * 32 + d2)>
```
@MaheshRavishankar, What do you think ?

**Comment by MaheshRavishankar** [comment]
I think right now there is a pattern in linalg fusion on tensors that folds reshape as consumer with its producer when it is a generic op, that does the linearization like you mentioned. At that time I though that is OK, but I have been wanting to take a look at changing that a little bit to avoid linearization. For example here, instead of fusing the tensor_reshape with its consumer, we can fuse it with the producer.

So take these two ops

```
%3 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = ["parallel", "parallel"]} ins(%0, %2 : tensor<384x128xf32>, tensor<384x128xf32>) {
    ^bb0(%arg0: f32, %arg1: f32):  // no predecessors
      %6 = addf %arg0, %arg1 : f32
      linalg.yield %6 : f32
    } -> tensor<384x128xf32>
%4 = linalg.tensor_reshape %3 [affine_map<(d0, d1, d2) -> (d0)>, affine_map<(d0, d1, d2) -> (d1, d2)>] : tensor<384x128xf32> into tensor<384x4x32xf32>
```

and fuse them to 

```
%3 = linalg.tensor_reshape %0 [affine_map<(d0, d1, d2) -> (d0)>, affine_map<(d0, d1, d2) -> (d1, d2)>] : tensor<384x128xf32> into tensor<384x4x32xf32>
%4 = linalg.tensor_reshape %2 affine_map<(d0, d1, d2) -> (d0)>, affine_map<(d0, d1, d2) -> (d1, d2)>] : tensor<384x128xf32> into tensor<384x4x32xf32>
%5 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel"]} ins(%0, %2 : tensor<384x4x32xf32>, tensor<384x4x32xf32>) {
    ^bb0(%arg0: f32, %arg1: f32):  // no predecessors
      %6 = addf %arg0, %arg1 : f32
      linalg.yield %6 : f32
    } -> tensor<384x4x32xf32>
```

Even though this is more ops the `tensor_reshape` produced would get fused with its producers, and bubble up to the top where it would get folded with the inputs. This would work for this example.

@hanhanW This is what I was mentioning to you about avoiding linearization.
So we can go with the linearization approach to make forward progress, but I was planning to implement this (which is needed when we try to fuse on buffers). I will take a look at this tonight to see if I can make forward progress on this fast. If it is taking me some time, we can use linearization (which will work within IREE) and I can go and adapt it later.

**Comment by MaheshRavishankar** [comment]
Actually pattern for this already seems to exist. Its just being conservative. Ill make the change and send it for review.

**Comment by MaheshRavishankar** [comment]
Looked at this, and I want to take this chance to make a more robust fusion of `tensor_reshape` ops for such cases. I dont think I will be able to push this in today (and I am going to be taking Friday off), but can definitely have this working by Monday. In the meantime if you want to unblock progress on this you (@asaadaldien ) can look into the linearization approach (it will still be useful, but as a fallback cause it the linearization approach only works for static shapes). To give an idea of what I have in mind here is a comment from my WIP change 

```
  // Consider                                                                                                                                                                                                                                                                                                     
  //                                                                                                                                                                                                                                                                                                              
  // %c = linalg.generic ins(%a, %b : memref<?x?x?xf32>, memref<?x?xf32>)                                                                                                                                                                                                                                         
  //        indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d0, d2)>,                                                                                                                                                                                                                                            
  //                         affine_map<(d0, d1, d2) -> (d1, d2)>,                                                                                                                                                                                                                                                
  //                         affine_map<(d0, d1, d2) -> (d0, d2, d1)>]                                                                                                                                                                                                                                            
  // %d = linalg.tensor_reshape %c                                                                                                                                                                                                                                                                                
  //        [affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d1)>,                                                                                                                                                                                                                                                    
  //         affine_map<(d0, d1, d2, d3, d4, d5) -> (d2)>,                                                                                                                                                                                                                                                        
  //         affine_map<(d0, d1, d2, d3, d4, d5) -> (d3, d4, d5)>]                                                                                                                                                                                                                                                
  //      : tensor<?x?x?xf32> into tensor<?x?x?x?x?x?xf32>                                                                                                                                                                                                                                                        
  //                                                                                                                                                                                                                                                                                                              
  // The reshape can be folded into the generic op if the generic op                                                                                                                                                                                                                                              
  // dimensionality is increased to match the result of the tensor_reshape. The                                                                                                                                                                                                                                   
  // indexing_map of the result and the reassociation map helps compute the                                                                                                                                                                                                                                       
  // indexing maps of the modified op. For the above example, based on the                                                                                                                                                                                                                                        
  // reassociation map it can be concluded that                                                                                                                                                                                                                                                                   
  //                                                                                                                                                                                                                                                                                                              
  // - The loop used to access the first dimension is split into two.                                                                                                                                                                                                                                             
  // - The loop used to access the second dimension is kept as is.                                                                                                                                                                                                                                                
  // - The loop used to access the third dimension is split into three.                                                                                                                                                                                                                                           
  //                                                                                                                                                                                                                                                                                                              
  // i.e. (e0, e1, e2, e3, e4) is the domain of the indexing map of the modified                                                                                                                                                                                                                                  
  // op, then                                                                                                                                                                                                                                                                                                     
  //                                                                                                                                                                                                                                                                                                              
  //  d0 -> e0, e1                                                                                                                                                                                                                                                                                                
  //  d1 -> e2, e3, e4                                                                                                                                                                                                                                                                                            
  //  d2 -> e5                                                                                                                                                                                                                                                                                                    
  //                                                                                                                                                                                                                                                                                                              
  // substituting this, the generic op can be rewritten as                                                                                                                                                                                                                                                        
  //                                                                                                                                                                                                                                                                                                              
  // %d = linalg.generic ins(%0, %1 : )                                                                                                                                                                                                                                                                           
  //         indexing_maps =                                                                                                                                                                                                                                                                                      
  //           [affine_map<(e0, e1, e2, e3, e4, e5) -> (e2, e3, e4, e0, e1, e5)>,                                                                                                                                                                                                                                 
  //            affine_map<(e0, e1, e2, e3, e4, e5) -> (e2, e3, e4, e5)>,                                                                                                                                                                                                                                         
  //            affine_map<(e0, e1, e2, e3, e4, e5) -> (e0, e1, e5, e2, e3, e4)>]                                                                                                                                                                                                                                 
  //                                                                                                                                                                                                                                                                                                              
  // Since operands to the linalg generic are now 5D, reshapes can be introduced                                                                                                                                                                                                                                  
  // to make it consistent                                                                                                                                                                                                                                                                                        
  //                                                                                                                                                                                                                                                                                                              
  // %0 = linalg.tensor_reshape %a                                                                                                                                                                                                                                                                                
  //        [affine_map<(e0, e1, e2, e3, e4, e5) -> (e0, e1, e2),                                                                                                                                                                                                                                                 
  //         affine_map<(e0, e1, e2, e3, e4, e5) -> (e3, e4),                                                                                                                                                                                                                                                     
  //         affine_map<(e0, e1, e2, e3, e4, e5) -> (e5)]                                                                                                                                                                                                                                                         
  //      : tensor<?x?x?xf32> into tensor<?x?x?x?x?x?xf32>                                                                                                                                                                                                                                                        
  // %1 = linalg.tensor_reshape %b                                                                                                                                                                                                                                                                                
  //        [affine_map<(e0, e1, e2, e3) -> (e0, e1, e2),                                                                                                                                                                                                                                                         
  //         affine_map<(e0, e1, e2, e3) -> (e3)]                                                                                                                                                                                                                                                                 
  //      : tensor<?x?x?xf32> into tensor<?x?x?x?xf32>                                                                                                                                                                                                                                                            
  //                                                                                                                                                                                                                                                                                                              
  // The added reshapes are again expanding patterns, so they will get fused                                                                                                                                                                                                                                      
  // with its producers if possible.                                                                                                                                                                                                                                                                         ````

**Comment by asaadaldien** [comment]
Thanks @MaheshRavishankar for Looking into this in details. Can you please point me to the linearization / folding pattern in Linalg?

**Comment by MaheshRavishankar** [comment]
THis is probably what you need : https://github.com/llvm/llvm-project/blob/9d40fb808fd0fbd33eb3b50c20d7f402de5db91e/mlir/lib/Dialect/Linalg/Transforms/FusionOnTensors.cpp#L431

Havent looked at it, but this is probably being very conservative.

**Comment by asaadaldien** [comment]
https://reviews.llvm.org/D88869 should fix this
